src/interactions/panZoomConstraints.ts

```diff
@@ -0,0 +1,176 @@
+/**
+ * Copyright 2017-present Palantir Technologies
+ * @license MIT
+ */
+
+import {TransformableScale} from "../scales/scale";
+
+/**
+ * Return `value` if its distance to `center` was scaled by `zoom`.
+ *
+ * e.g. zoomOut(100, 2, 50) -> 150
+ * e.g. zoomOut(0, 2, 50) -> -50
+ * e.g. zoomOut(100, 0.5, 50) -> 75
+ * e.g. zoomOut(0, 0.5, 50) -> 25
+ */
+export function zoomOut(value: number, zoom: number, center: number) {
+  return center - (center - value) * zoom;
+}
+
+/**
+ * This is the zoomOut method algebra-ed to solve for the `center` value
+ * given the other three. The "target" is the return value of zoomOut.
+ */
+function getZoomOutCenter(value: number, zoom: number, target: number) {
+  return (value * zoom - target) / (zoom - 1);
+}
+
+/**
+ * Return possibly different zoomAmount and centerPoint values such that applying the zoom
+ * to the scale respects the given min/max extents and values.
+ */
+export function constrainedZoom(
+  scale: TransformableScale<any, number>,
+  zoomAmount: number,
+  centerPoint: number,
+  minDomainExtent: number,
+  maxDomainExtent: number,
+  minDomainValue: number,
+  maxDomainValue: number,
+  ) {
+  zoomAmount = constrainZoomExtents(scale, zoomAmount, minDomainExtent, maxDomainExtent);
+  return constrainZoomValues(scale, zoomAmount, centerPoint, minDomainValue, maxDomainValue);
+}
+
+function constrainZoomExtents(
+  scale: TransformableScale<any, number>,
+  zoomAmount: number,
+  minDomainExtent: number,
+  maxDomainExtent: number,
+) {
+  const extentIncreasing = zoomAmount > 1;
+
+  const boundingDomainExtent = extentIncreasing ? maxDomainExtent : minDomainExtent;
+  if (boundingDomainExtent == null) {
+    return zoomAmount;
+  }
+
+  const [ scaleDomainMin, scaleDomainMax ] = scale.getTransformationDomain();
+  const domainExtent = Math.abs(scaleDomainMax - scaleDomainMin);
+  const compareF = extentIncreasing ? Math.min : Math.max;
+  return compareF(zoomAmount, boundingDomainExtent / domainExtent);
+}
+
+/**
+ * Modify zoomAmount and centerPoint such that a panzoom will stay inside
+ * the bounds defined by minDomainValue and maxDomainValue.
+ */
+function constrainZoomValues(
+  scale: TransformableScale<any, number>,
+  zoomAmount: number,
+  centerPoint: number,
+  minDomainValue: number,
+  maxDomainValue: number,
+  ) {
+  // when zooming in, we don't have to worry about overflowing domain
+  if (zoomAmount <= 1) {
+    return { centerPoint, zoomAmount };
+  }
+
+  const reversed = isRangeReversed(scale);
+  // if no constraints set, we're done
+  if (minDomainValue == null && maxDomainValue == null) {
+    return { centerPoint, zoomAmount };
+  }
+
+  minDomainValue = minDomainValue == null ? -Infinity : minDomainValue;
```

Comment:
won't `-Infinity` be wrong if the domain is reversed? maybe verify this by removing the max constraint from the pan_zoom_reverse preview

Reply:
otherwise, i like this approach to clean up this function